io/dagreader.go

```diff
@@ -66,12 +76,421 @@ func NewDagReader(ctx context.Context, n ipld.Node, serv ipld.NodeGetter) (DagRe
+				fsNode, err := unixfs.ExtractFSNode(node)
+				if err != nil {
+					return err
+				}
+
+				// If there aren't enough size hints don't seek
+				// (see the `io.EOF` handling error comment below).
+				if fsNode.NumChildren() != len(node.Links()) {
+					return io.EOF
+				}
+
+				// Internal nodes have no data, so just iterate through the
+				// sizes of its children (advancing the child index of the
+				// `dagWalker`) to find where we need to go down to next in
+				// the search.
+				for {
+					childSize := fsNode.BlockSize(int(dr.dagWalker.ActiveChildIndex()))
+
+					if childSize > uint64(left) {
+						// This child's data contains the position requested
+						// in `offset`, go down this child.
+						return nil
+					}
+
+					// Else, skip this child.
+					left -= int64(childSize)
+					err := dr.dagWalker.NextChild()
+					if err == ipld.ErrNextNoChild {
+						// No more child nodes available, nothing to do,
+						// the `Seek` will stop on its own.
+						return nil
+					} else if err != nil {
+						return err
+						// Pass along any other errors (that may in future
+						// implementations be returned by `Next`) to stop
+						// the search.
+					}
+				}
+
+			} else {
+				// Leaf node, seek inside its data.
+				err := dr.saveNodeData(node)
+				if err != nil {
+					return err
+				}
+
+				_, err = dr.currentNodeData.Seek(left, io.SeekStart)
+				if err != nil {
+					return err
+				}
+				// The corner case of a DAG consisting only of a single (leaf)
+				// node should make no difference here. In that case, where the
+				// node doesn't have a parent UnixFS node with size hints, this
+				// implementation would allow this `Seek` to be called with an
+				// argument larger than the buffer size which normally wouldn't
+				// happen (because we would skip the node based on the size
+				// hint) but that would just mean that a future `CtxReadFull`
+				// call would read no data from the `currentNodeData` buffer.
+				// TODO: Re-check this reasoning.
+
+				return nil
+				// In the leaf node case the search will stop here.
+			}
+		})
+
+		if err == io.EOF {
+			// TODO: Taken from https://github.com/ipfs/go-ipfs/pull/4320,
+			// check if still valid.
+			//   Return negative number if we can't figure out the file size. Using io.EOF
+			//   for this seems to be good(-enough) solution as it's only returned by
+			//   precalcNextBuf when we step out of file range.
+			//   This is needed for gateway to function properly
+			return -1, nil
+		}
+
+		if err != nil {
+			return 0, err
+		}
+
+		dr.offset = offset
+		return dr.offset, nil
+
+	case io.SeekCurrent:
+		if offset == 0 {
+			return dr.offset, nil
+		}
+
+		return dr.Seek(dr.offset+offset, io.SeekStart)
+		// TODO: Performance. This can be improved supporting relative
+		// searches in the `Walker` (see `Walker.Seek`).
+
+	case io.SeekEnd:
+		return dr.Seek(int64(dr.Size())-offset, io.SeekStart)
```

Comment:
```suggestion
		return dr.Seek(int64(dr.Size())+offset, io.SeekStart)
```

This is also a bug in current implementation - https://golang.org/pkg/io/#Seeker - https://golang.org/src/bytes/reader.go?s=2744:2806#L103 - https://github.com/ipfs/go-ipfs/pull/5934